crypto/cmp/cmp_vfy.c

```diff
@@ -0,0 +1,868 @@
+    }
+    X509_STORE_free(store);
+    return valid;
+}
+
+/* Find and validate any potentially usable server cert */
+static X509 *find_srvcert(OSSL_CMP_CTX *ctx, const OSSL_CMP_MSG *msg)
+{
+    X509 *scrt = NULL;
+    int valid = 0;
+    GENERAL_NAME *sender = msg->header->sender;
+
+    if (sender == NULL || msg->body == NULL)
+        return 0; /* other NULL cases already have been checked */
+    if (sender->type != GEN_DIRNAME) {
+        CMPerr(0, CMP_R_SENDER_GENERALNAME_TYPE_NOT_SUPPORTED);
+        return NULL; /* FR#42: support for more than X509_NAME */
+    }
+
+    /*
+     * valid scrt, matching sender name, found earlier in transaction, will be
+     * used for validating any further msgs where extraCerts may be left out
+     */
+    (void)ERR_set_mark();
+    if (ctx->validatedSrvCert != NULL
+            && cert_acceptable(ctx->validatedSrvCert, msg, ctx->trusted)) {
+        (void)ERR_pop_to_mark();
+        scrt = ctx->validatedSrvCert;
+        valid = 1;
+    } else {
+        STACK_OF(X509) *found_crts;
+        char *sname;
+        int i;
+
+        (void)ERR_pop_to_mark();
+
+        /* sk_TYPE_find to use compfunc X509_cmp, not ptr comparison */
+        if ((found_crts = sk_X509_new_null()) == NULL)
+            return NULL;
+
+        /* tentatively set error, which allows accumulating diagnostic info */
+        (void)ERR_set_mark();
+        CMPerr(0, CMP_R_NO_VALID_SERVER_CERT_FOUND);
+        sname = X509_NAME_oneline(sender->d.directoryName, NULL, 0);
+        ossl_cmp_add_error_txt(NULL, "\n");
+        ossl_cmp_add_error_txt("trying to match msg sender name = ", sname);
+        OPENSSL_free(sname);
+
+        /* release any cached cert, which is no more acceptable */
+        (void)ossl_cmp_ctx_set0_validatedSrvCert(ctx, NULL);
+
+        /* find server cert candidates from any available source */
+        if (!find_server_cert(ctx->trusted, ctx->untrusted_certs,
+                              msg, found_crts))
+            goto end;
+
+        /* select first server cert that can be validated */
+        for (i = 0; !valid && i < sk_X509_num(found_crts); i++) {
+            scrt = sk_X509_value(found_crts, i);
+            valid = OSSL_CMP_validate_cert_path(ctx, ctx->trusted, scrt);
+        }
+
+        /* exceptional 3GPP TS 33.310 handling */
+        if (!valid && ctx->permitTAInExtraCertsForIR
+                && ossl_cmp_msg_get_bodytype(msg) == OSSL_CMP_PKIBODY_IP) {
+            for (i = 0; !valid && i < sk_X509_num(found_crts); i++) {
+                scrt = sk_X509_value(found_crts, i);
+                valid = srv_cert_valid_3gpp(ctx, scrt, msg);
+            }
+        }
+
+        if (valid) {
+            /* discard any diagnostic info on finding server cert */
+            (void)ERR_pop_to_mark();
+            /* store trusted srv cert for future msgs of same transaction */
+            if (ossl_cmp_ctx_set0_validatedSrvCert(ctx, scrt))
+                X509_up_ref(scrt);
```

Comment:
up_ref first. Check for success, then call set0.

Reply:
Thanks, done.